pinning/pinner/dspinner/pin.go

```diff
@@ -253,7 +287,17 @@ func (p *pinner) doPinRecursive(ctx context.Context, c cid.Cid, fetch bool, name
 	if err != nil {
 		return err
 	}
-	return p.flushPins(ctx, false)
+	err = p.flushPins(ctx, false)
+	if err != nil {
+		return err
+	}
+
+	// Provide only if we have not set pinnedProvider, as otherwise
+	// we would provide it twice.
+	if p.rootsProvider != nil && p.pinnedProvider == nil {
+		return p.rootsProvider.Provide(ctx, c, true)
```

Comment:
```suggestion
		go p.rootsProvider.Provide(ctx, c, true)
```

unless we want the pin operation to fail on provide failure.